BRASS
=====

### Breakpoints via assembly

BRASS analyses one or more related BAM files of paired-end sequencing to determine potential rearrangement breakpoints.

There are several stages the main component being:

1. Collect read-pairs where both ends map but **NOT** marked as properly-paired.
2. Perform grouping based on mapped locations
3. Filter
4. Run assembly
5. Annotate with [GRASS](http://cancerit.github.io/grass/)

---

## Installation

    ./setup.sh path_to_install_to

### Pre-requisites

* The C++ code (within this package) requires the presence of pstreams.h (and associated development libraries).  This is not handled by the `setup.sh` script.

##### Perl packages:

* [PCAP-core 2.1.3+](https://github.com/ICGC-TCGA-PanCancer/PCAP-core/releases)
* [cgpVcf 2.0.1+](https://github.com/cancerit/cgpVcf/releases)
* [grass 2.0.2+](https://github.com/cancerit/grass/releases)

##### R packages

A large number of R packages are required to run BRASS.  To facilitate the install process there is a script `Rsupport/libInstall.R` that can be run to build these for you.  See this file for the list of packages.

Alternatively you can run:

````
cd Rsupport
./setupR.sh path_to_install_to
````

Appending `1` to the command will request a complete local build of `R` (3.1.3).


##### Other tools that need to be in path

* [FASTA](https://github.com/wrpearson/fasta36/releases)

###  Tools installed by setup.sh
* Many CPAN hosted libraries, see `perl/Makefile.PL` for the list
* [BLAT](http://hgwdev.cse.ucsc.edu/~kent/src/)
* [exonerate](http://www.ebi.ac.uk/about/vertebrate-genomics/software/exonerate)
* [Velvet](https://www.ebi.ac.uk/~zerbino/velvet/)
* [bedtools2](https://github.com/arq5x/bedtools2/releases)

Please use `setup.sh` to install these dependencies.  Setting the environment variable `CGP_PERLLIBS` allows you to to append to `PERL5LIB` during install.  Without this all dependancies are installed into the target area.  `setup.sh` will not use `PERL5LIB` directly.

Please be aware that this expects basic C compilation libraries and tools to be available.

###

## Running BRASS

This package includes a reference implementation which handles all of the linking together of steps.

Please see the ``-h`` and ``-m`` options of ``brass.pl`` for full usage information.

It can be run in a couple of ways:

1. Fire and forget
  * Execute on a single host with multiple cores (or 1 if that's all you have)
  * Some efficiency overhead as some steps aren't parallel
2. Farm style
  * Requires 2 extra parameters in the initial command
  * See ``-help`` for further details

---

## Input

### Initial mapping

BRASS has primarily been written to work with BWA mapped data.  You are likely to get the most useful output from BWA-mem.

### Library quality

Please be aware that paired-end libraries where properly-paired reads overlap are unlikely to produce good results.

### Additional mapping information

BRASS requires accurate information regarding the insert size distribution and expects to find a ``*.bam.bas`` file
co-located with the ``*.bam``'s.  These can be generated by the ``bam_stats`` program included in the
[PCAP-core](https://github.com/ICGC-TCGA-PanCancer/PCAP-core) project.  If you use ``bwa_mem.pl`` to map your
data (same repository) then this file is generated automatically for you.

LICENCE
=======
Copyright (c) 2014-2016 Genome Research Ltd.

Author: Cancer Genome Project <cgpit@sanger.ac.uk>

This file is part of BRASS.

BRASS is free software: you can redistribute it and/or modify it under
the terms of the GNU Affero General Public License as published by the Free
Software Foundation; either version 3 of the License, or (at your option) any
later version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
details.

You should have received a copy of the GNU Affero General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.

1. The usage of a range of years within a copyright statement contained within
this distribution should be interpreted as being equivalent to a list of years
including the first and last year specified and all consecutive years between
them. For example, a copyright statement that reads ‘Copyright (c) 2005, 2007-
2009, 2011-2012’ should be interpreted as being identical to a statement that
reads ‘Copyright (c) 2005, 2007, 2008, 2009, 2011, 2012’ and a copyright
statement that reads ‘Copyright (c) 2005-2012’ should be interpreted as being
identical to a statement that reads ‘Copyright (c) 2005, 2006, 2007, 2008,
2009, 2010, 2011, 2012’."